rend3-routine/src/skinning/mod.rs

```diff
@@ -0,0 +1,272 @@
+        // skinning input struct contains an offset into this array.
+        let joint_matrices_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(true), None);
+            bglb.build(device, Some("Gpu skinning joint matrices"))
+        };
+
+        // Bind group 2 contains the pre skinning inputs. This uses dynamic
+        // offsets because there is one dispatch per input, and the offset is
+        // used to indicate which is the current input to the shader.
+        //
+        // NOTE: This would be an ideal use case for push constants, but they are
+        // not available on all platforms so we need to use this workaround.
+        let skinning_inputs_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(
+                wgpu::ShaderStages::COMPUTE,
+                wgpu::BindingType::Buffer {
+                    ty: wgpu::BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: None,
+                },
+                None,
+            );
+            bglb.build(device, Some("Gpu skinning inputs"))
+        };
+
+        let layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
+            label: None,
+            bind_group_layouts: &[&vertex_buffers_bgl, &joint_matrices_bgl, &skinning_inputs_bgl],
+            push_constant_ranges: &[],
+        });
+
+        let module = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
+            label: Some("Gpu skinning compute shader"),
+            source: wgpu::ShaderSource::Wgsl(include_str!("../../shaders/src/skinning.wgsl").into()),
+        });
+
+        let pipeline = device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
+            label: Some("Gpu skinning pipeline"),
+            layout: Some(&layout),
+            module: &module,
+            entry_point: "main",
+        });
+
+        GpuSkinner {
+            vertex_buffers_bgl,
+            joint_matrices_bgl,
+            skinning_inputs_bgl,
+            pipeline,
+        }
+    }
+
+    pub fn execute_pass(
+        &self,
+        device: &Device,
+        encoder: &mut wgpu::CommandEncoder,
+        buffers: &PreSkinningBuffers,
+        mesh_buffers: &MeshBuffers,
+        // The number of inputs in the skinning_inputs buffer
+        skeleton_manager: &SkeletonManager,
+    ) {
+        let vertex_buffers_bg = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&mesh_buffers.vertex_position);
+            bgb.append_buffer(&mesh_buffers.vertex_normal);
+            bgb.append_buffer(&mesh_buffers.vertex_tangent);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_index);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_weight);
+            bgb.build(device, Some("GPU skinning vertex buffers"), &self.vertex_buffers_bgl)
```

Comment:
Super duper nit: these can be chained, so `BindGroupBuilder::new().append_buffer(...).build()` and we don't need the mut or the block here

Reply:
Much nicer, thanks! :smile: